REGRESSION(257219@main): [GTK] Cursor blinks oddly in text fields, even if blinking disabled

aperezdc opened the pull request:
#### 6a0a83593474f3c4c6959c4d36de25315094b643
<pre>
REGRESSION(257219@main): [GTK] Cursor blinks oddly in text fields, even if blinking disabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=252687">https://bugs.webkit.org/show_bug.cgi?id=252687</a>

Reviewed by Carlos Garcia Campos.

Before 257219@main if RenderTheme::caretBlinkInterval() returned zero
then the timer controlling the blink animation was never started. When
introducing CaretAnimator + SimpleCaretAnimator the check was not
carried over, resulting in the animator trying to paint the caret
on-and-off as fast as possible when blinking should have been disabled.

The fix reintroduces the check, reusing setBlinkingSuspended() to enable
(or disable) blinking depending on RenderTheme::caretBlinkInterval(),
which in turn is changed to return std::optional&lt;Seconds&gt;, with
std::nullopt signalling that blinking is disables. The improved typing
ensures that the code explicitly handles the case of blinking being
disabled.

While at it, remove the RenderThemeGtk.{h,cpp} files, because they only
contained an implementation of caretBlinkInterval() that was exactly the
same code guarded with PLATFORM(GTK) in RenderThemeAdwaita.

* Source/WebCore/SourcesGTK.txt: Remove RenderThemeGtk.cpp from the list.
* Source/WebCore/platform/CaretAnimator.h:
(WebCore::CaretAnimator::didStart): Adapt to use std::optional&lt;Seconds&gt;.
* Source/WebCore/platform/SimpleCaretAnimator.cpp:
(WebCore::SimpleCaretAnimator::updateAnimationProperties): Arrange to
enable or disable caret blinking, adapting to use std::optional&lt;Seconds&gt;.
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::caretBlinkInterval const): Change the return type
to std::optional&lt;Seconds&gt;.
* Source/WebCore/rendering/RenderThemeAdwaita.cpp:
(WebCore::RenderTheme::singleton): Remove PLATFORM(GTK) guard.
(WebCore::RenderThemeAdwaita::caretBlinkInterval const): Adapt to use
std::optional&lt;Seconds&gt;.
* Source/WebCore/rendering/RenderThemeAdwaita.h: Ditto.
* Source/WebCore/rendering/RenderThemeGtk.cpp: Removed.
* Source/WebCore/rendering/RenderThemeGtk.h: Removed.

Canonical link: <a href="https://commits.webkit.org/262954@main">https://commits.webkit.org/262954@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/55fc48da9609c3f778a7d1f972cd0d1bae2d0b71

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3163 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3224 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3334 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4576 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3521 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3131 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3285 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3251 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/2752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3194 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/3510 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/2842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4383 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/1008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/2816 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/2656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/2787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/2867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/4131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3239 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2823 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/2818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2816 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/366 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3082 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/rendering/RenderThemeAdwaita.cpp:
```diff
@@ -695,12 +693,14 @@ void RenderThemeAdwaita::adjustListButtonStyle(RenderStyle& style, const Element
-    return shouldBlink ? 500_us * time : 0_s;
+    if (shouldBlink)
+        return { 500_us * time };
+    return { };
```
Shouldn't this be std::nullopt?

aperezdc replied:
Empty initializer list is equivalent to `std::nullopt`, I have seen both styles in WebKit. I can change it before landing if you prefer.

Ms2ger commented on Source/WebCore/platform/SimpleCaretAnimator.cpp:
```diff
@@ -39,17 +39,23 @@ void SimpleCaretAnimator::updateAnimationProperties(ReducedResolutionSeconds cur
     }
 
+    // If blinking is disabled, set isBlinkingSuspended() would have made the
+    // previuos check return early and at this point there must be an interval.
```
```suggestion
    // previous check return early and at this point there must be an interval.
```

aperezdc replied:
Thanks, I will fix the typo before landing 👌🏼

aperezdc committed:
REGRESSION(257219@main): [GTK] Cursor blinks oddly in text fields, even if blinking disabled
https://bugs.webkit.org/show_bug.cgi?id=252687

Reviewed by Carlos Garcia Campos.

Before 257219@main if RenderTheme::caretBlinkInterval() returned zero
then the timer controlling the blink animation was never started. When
introducing CaretAnimator + SimpleCaretAnimator the check was not
carried over, resulting in the animator trying to paint the caret
on-and-off as fast as possible when blinking should have been disabled.

The fix reintroduces the check, reusing setBlinkingSuspended() to enable
(or disable) blinking depending on RenderTheme::caretBlinkInterval(),
which in turn is changed to return std::optional<Seconds>, with
std::nullopt signalling that blinking is disables. The improved typing
ensures that the code explicitly handles the case of blinking being
disabled.

While at it, remove the RenderThemeGtk.{h,cpp} files, because they only
contained an implementation of caretBlinkInterval() that was exactly the
same code guarded with PLATFORM(GTK) in RenderThemeAdwaita.

* Source/WebCore/SourcesGTK.txt: Remove RenderThemeGtk.cpp from the list.
* Source/WebCore/platform/CaretAnimator.h:
(WebCore::CaretAnimator::didStart): Adapt to use std::optional<Seconds>.
* Source/WebCore/platform/SimpleCaretAnimator.cpp:
(WebCore::SimpleCaretAnimator::updateAnimationProperties): Arrange to
enable or disable caret blinking, adapting to use std::optional<Seconds>.
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::caretBlinkInterval const): Change the return type
to std::optional<Seconds>.
* Source/WebCore/rendering/RenderThemeAdwaita.cpp:
(WebCore::RenderTheme::singleton): Remove PLATFORM(GTK) guard.
(WebCore::RenderThemeAdwaita::caretBlinkInterval const): Adapt to use
std::optional<Seconds>.
* Source/WebCore/rendering/RenderThemeAdwaita.h: Ditto.
* Source/WebCore/rendering/RenderThemeGtk.cpp: Removed.
* Source/WebCore/rendering/RenderThemeGtk.h: Removed.

Canonical link: https://commits.webkit.org/262954@main
